File: boj15686_치킨배달.java
import java.io.*;
import java.util.*;
import java.math.*;

class boj15686_치킨배달 {
    static int n;
    static int m;
    static int[][] board;
    static boolean[] visited;
    static ArrayList<Coord> homes;
    static ArrayList<Coord> stores;
    static ArrayList<Integer> comb;
    static int ans;

    public static class Coord {
        public int x;
        public int y;

        public Coord(int _x, int _y) {
            x = _x;
            y = _y;
        }
    }

    public static void initArgs() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        board = new int[n][n];

        for (int i = 0; i < n; ++i) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; ++j) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        homes = new ArrayList();
        stores = new ArrayList();
        comb = new ArrayList();

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 1)
                    homes.add(new Coord(i, j));
                if (board[i][j] == 2)
                    stores.add(new Coord(i, j));
            }
        }

        visited = new boolean[stores.size()];
        ans = Integer.MAX_VALUE;
    }

    public static void solve() {
        int chickenDist = 0;

        for (int i = 0; i < homes.size(); ++i) {
            int minDist = Integer.MAX_VALUE;
            for (int j = 0; j < comb.size(); ++j) {
                int idx = comb.get(j);
                int dist = Math.abs(homes.get(i).x - stores.get(idx).x) + Math.abs(homes.get(i).y - stores.get(idx).y);
                minDist = Math.min(minDist, dist);
            }
            chickenDist += minDist;
        }
        ans = Math.min(chickenDist, ans);
    }

    public static void dfs(int depth) {
        if (depth >= m) {
            // System.out.println(comb);
            solve();
            return ;
        }

        for (int i = 0; i < stores.size(); ++i) {
            if (visited[i]) {
                continue;
            }
            comb.add(i);
            visited[i] = true;
            dfs(depth + 1);
            comb.remove(comb.size() - 1);
            for (int j = i+1; j < stores.size(); ++j) {
                visited[j] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        initArgs();
        dfs(0);
        System.out.println(ans);
    }
}

File: boj15686-치킨배달.kt
import kotlin.math.abs
import kotlin.math.min

class Boj15686 {
    val n: Int
    val m: Int
    val city: Array<IntArray>
    var ans = Int.MAX_VALUE
    val homes = mutableListOf<Pair<Int,Int>>()
    val chickens = mutableListOf<Pair<Int, Int>>()
    val lst = mutableListOf<Int>()

    init {
        readln().split(' ').map { it.toInt() }.also {
            n=it[0]
            m=it[1]
        }
        city = Array(n) { readln().split(' ').map { it.toInt() }.toIntArray() }
        for (i in city.indices) {
            for (j in city[i].indices) {
                if (city[i][j] == 1)
                    homes += Pair(i, j)
                if (city[i][j] == 2)
                    chickens += Pair(i, j)
            }
        }
    }

    fun checkChickenDist(visit: BooleanArray) {
        var chickenDist = 0
        for (i in homes.indices) {
            var dist = Int.MAX_VALUE
            for (j in chickens.indices) {
                if (j !in lst)
                    continue
                val tmp = abs(homes[i].first - chickens[j].first) + abs(homes[i].second - chickens[j].second)
                if (dist > tmp)
                    dist = tmp
            }
            if (dist != Int.MAX_VALUE)
                chickenDist += dist
        }
        ans = min(chickenDist, ans)
    }

    fun dfs(cnt: Int, visit: BooleanArray) {
        if (cnt == m) {
            checkChickenDist(visit)
            return
        }

        for (i in chickens.indices) {
            if (visit[i])
                continue
            visit[i] = true
            lst += i
            dfs(cnt + 1, visit)
            lst.removeLast()
            for (j in i + 1 until visit.size)
                visit[j] = false
        }
    }

    fun solution() {
        dfs(0, BooleanArray(chickens.size) {false})
        println(ans)
    }
}

fun main() {
    Boj15686().solution()
}